### PR TITLE
Fixes for CG filesets

### PR DESCRIPTION
**Pull request by amdabhad**
## Pull request checklist

Fixes in CG fileset discovery for version 2 volumes:
* Change in filter to filter only independent filesets
* Handle paginated response from GUI (for more than 1000 filesets GUI uses pagination in response)

## Pull request type

Please check the type of change your PR introduces:
- [x] Bugfix
- [ ] Feature Enhancement
- [ ] Test Automation 
- [ ] Code Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Community Operator listing 
- [ ] Other (please describe): 

## What is the current behavior?
All filesets are filtered for CG fileset discovery, paginated response from GUI is not handled

## What is the new behavior?
Only independent filesets are filtered for CG fileset discovery and paginated response from GUI is handled

## How risky is this change?
- [x] Small, isolated change
- [ ] Medium, requires regression testing
- [ ] Large, requires functional and regression testing

### CODE REVIEW

File: driver/csiplugin/connectors/rest_v2.go [hunk 1]
```diff
@@ -718,7 +721,33 @@ func (s *SpectrumRestV2) ListCSIIndependentFilesets(ctx context.Context, filesys
...
+		filesets = append(filesets, getFilesetResponse.Filesets...)
+	}
+	return filesets, nil
+}
+
+func getLastID(nextURL string) string {
+	//Sample next URL format:
+	//"/scalemgmt/v2/filesystems/fs2/filesets?lastId=4368&fields=filesetName&filter=config.isInodeSpaceOwner=true,config.comment=Fileset created by IBM Container Storage Interface driver"
+	splitURL1 := strings.Split(nextURL, "&")
+	splitURL2 := strings.Split(splitURL1[0], "lastId=")
```
Comment (deeghuge) [inline driver/csiplugin/connectors/rest_v2.go:747]: @amdabhad do confirm with gui team if the sequence of lastId and fields always remain in this sequence.

Comment (amdabhad) [reply driver/csiplugin/connectors/rest_v2.go:747]: Confirmed with GUI, the sequence remains the same

**Comment by amdabhad** [comment]
Driver image: quay.io/amdabhad/csi-driver:cgFilesetFixes1001

**Review by deeghuge — APPROVED** [review]

File: driver/csiplugin/connectors/rest_v2.go [hunk 2]
```diff
@@ -718,7 +721,33 @@ func (s *SpectrumRestV2) ListCSIIndependentFilesets(ctx context.Context, filesys
 		return nil, err
 	}
 
-	return getFilesetResponse.Filesets, nil
+	filesets := getFilesetResponse.Filesets
+
+	emptyPages := Pages{}
+	for getFilesetResponse.Paging != emptyPages {
+		lastID := getLastID(getFilesetResponse.Paging.Next)
```
Comment (Jainbrt) [inline driver/csiplugin/connectors/rest_v2.go:728]: I feel we can get the lastId using getFilesetResponse.Paging.LastID directly ?

  "paging" : {
    "baseUrl" : "/scalemgmt/v2/filesystems/local-sample/filesets",
    "fields" : "filesetName",
    "lastId" : 1245,
    "next" : "/scalemgmt/v2/filesystems/local-sample/filesets?lastId=1245&fields=filesetName"
  },

File: driver/csiplugin/connectors/rest_v2.go [hunk 3]
```diff
@@ -709,7 +709,10 @@ func (s *SpectrumRestV2) ListCSIIndependentFilesets(ctx context.Context, filesys
 	klog.V(4).Infof("[%s] rest_v2 ListCSIIndependentFilesets. filesystem: %s", utils.GetLoggerId(ctx), filesystemName)
 
 	encodedFilesetComment := strings.ReplaceAll(FilesetComment, " ", "%20")
-	getFilesetURL := fmt.Sprintf("scalemgmt/v2/filesystems/%s/filesets?fields=filesetName&filter=config.parentId=0,config.comment=%s", filesystemName, encodedFilesetComment)
+	url := fmt.Sprintf("scalemgmt/v2/filesystems/%s/filesets?fields=filesetName", filesystemName)
+	filter := fmt.Sprintf("filter=config.isInodeSpaceOwner=true,config.comment=%s", encodedFilesetComment)
```
Comment (Jainbrt) [inline driver/csiplugin/connectors/rest_v2.go:713]: Also to filter independent filesets, we should filter on parentId=0.

Comment (amdabhad) [reply driver/csiplugin/connectors/rest_v2.go:713]: The dependent filesets can also have parentId=0 if they use the inode space of root fileset, so we can't use this for filter

File: driver/csiplugin/connectors/rest_v2.go (hunk 2 above)
Comment (amdabhad) [reply driver/csiplugin/connectors/rest_v2.go]: The struct Pages in CSI had only next field earlier so CSI was not reading other fields, have added other fields now and improved this part

**Comment by amdabhad** [comment]
New driver image: quay.io/amdabhad/csi-driver:CGFilesetFix1030

**Review by Jainbrt — DISMISSED** [review]

**Review by Jainbrt — APPROVED** [review]

**Review by deeghuge — APPROVED** [review]

**Pull request merged**